File: shared/tooling/qa-tool.mjs
// The launcher every skill invokes to reach the deterministic engine.
//
// A skill's SKILL.md documents exactly one command shape:
//
//     node <SKILL_DIR>/scripts/qa-tool.mjs <tool> <subcommand> [args]
//
// and this file finds the engine, wherever it happens to be. That indirection
// exists because the pack is installed three different ways and the engine lands in
// a different place each time:
//
//   1. `qa install` bundles the engine into the skill, at ./lib/. Offline, fastest,
//      and pinned to the version that was installed.
//   2. `npx skills add <owner>/<repo>` — or any generic file copier — copies the
//      skill directory out of git and bundles nothing. The engine is then resolved
//      from node_modules if the project happens to depend on the pack.
//   3. Neither: fall back to `npx qa-engineer`, which fetches the published package
//      on first use and is served from the npm cache afterwards.
//
// Because this file is committed rather than generated, path 2 works at all — which
// is what makes the pack installable by the wider Agent Skills ecosystem. The
// command a skill runs never changes; only where the engine came from does.
//
// Exit codes pass through unchanged: 0 success, 1 an invalid contract, 2 unreadable
// input or bad usage.
//
// No shebang: the synced copies carry a provenance marker on line one, which would
// sit above it and stop the kernel seeing it anyway. Every documented invocation is
// `node qa-tool.mjs …`, which needs none.

import fs from 'node:fs';
import path from 'node:path';
import { spawnSync } from 'node:child_process';
import { fileURLToPath } from 'node:url';

const here = path.dirname(fileURLToPath(import.meta.url));

const USAGE = `usage: node qa-tool.mjs <tool> <subcommand> [args]

  analysis     parse artifacts, classify errors, validate contracts, diff-guard,
               read .qa/context.md, render an HTML report, print the footer
  diagnostics  root cause, timeline, priority, repair plans, release readiness
  playwright   normalize a Playwright report or summarize a trace

  --where      print how the engine was resolved, and stop

examples:
  node qa-tool.mjs analysis junit test-results/results.xml
  node qa-tool.mjs analysis report-html qa-artifacts/explore-result.json --out report.html
  node qa-tool.mjs diagnostics report --execution-result qa-artifacts/run.json
`;

/**
 * Where the engine is, and how we found it.
 *
 * Ordered by cost: a bundled copy needs no resolution and no network, a
 * node_modules copy needs no network, and npx needs both on first use. Reporting
 * *which* one answered matters when a skill degrades — "the engine is missing" and
 * "the engine is being fetched" are different problems.
 */
function resolveEngine() {
  const bundled = path.join(here, 'lib', 'bin', 'qa-engine.mjs');
  if (fs.existsSync(bundled)) return { kind: 'bundled', command: [process.execPath, bundled] };

  for (const base of candidateModuleRoots()) {
    const installed = path.join(base, 'qa-engineer', 'packages', 'engine', 'bin', 'qa-engine.mjs');
    if (fs.existsSync(installed)) {
      return { kind: 'node_modules', command: [process.execPath, installed] };
    }
  }

  return {
    kind: 'npx',
    command: [npxCommand(), '--yes', 'qa-engineer', 'engine'],
  };
}

/** node_modules directories worth checking, nearest first. */
function candidateModuleRoots() {
  const roots = [];
  let dir = here;
  for (let depth = 0; depth < 12; depth += 1) {
    roots.push(path.join(dir, 'node_modules'));
    const parent = path.dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  if (process.cwd() !== here) roots.push(path.join(process.cwd(), 'node_modules'));
  return roots;
}

// `npx` is a shell script on POSIX and a .cmd shim on Windows; spawnSync needs the
// exact name, and `shell: true` would put user-supplied arguments through a shell.
function npxCommand() {
  return process.platform === 'win32' ? 'npx.cmd' : 'npx';
}

function main(argv) {
  if (argv.length === 0 || argv[0] === '--help' || argv[0] === '-h') {
    process.stdout.write(USAGE);
    return argv.length === 0 ? 2 : 0;
  }

  const engine = resolveEngine();

  if (argv[0] === '--where') {
    process.stdout.write(`${JSON.stringify({ resolved: engine.kind, command: engine.command }, null, 2)}\n`);
    return 0;
  }

  const [program, ...prefix] = engine.command;
  const run = spawnSync(program, [...prefix, ...argv], { stdio: 'inherit' });

  if (run.error) {
    // Say which path was tried and what to do, because a skill's fallback prose
    // cannot diagnose this and the user is the one who has to fix it.
    const advice = engine.kind === 'npx'
      ? 'the engine is not bundled and npx is unavailable — run `npx qa-engineer install` in this project, or install Node 18+'
      : `could not execute ${program}`;
    process.stderr.write(`qa-tool: ${advice}\n${run.error.message}\n`);
    return 2;
  }
  return run.status ?? 2;
}

process.exitCode = main(process.argv.slice(2));

File: shared/tooling/qa_tool.py
#!/usr/bin/env python3
"""The single entry point for a skill's bundled deterministic tooling.

Every skill that carries an engine also carries a copy of this file at
`<skill>/scripts/qa_tool.py`, beside the `lib/` directory it dispatches into.

## Why this exists

The invocation used to be a shell recipe:

    QA_LIB="$(ls -d .agents/skills/qa-run/scripts/lib ... | head -1)"
    PYTHONPATH="$QA_LIB" python3 -m qa_analysis.cli junit report.xml

Three parts of that are POSIX-only — command substitution, `ls -d | head -1`, and
the `VAR=value command` prefix — so on Windows PowerShell every deterministic call
failed. Failing calls do not stop a skill; they push it onto its documented manual
fallback. The result was that Windows users silently got the *guessing* behaviour
this project exists to replace, while believing they had the tooling.

So the shell does no work at all now. This launcher finds its own `lib/`
directory from `__file__` and puts it on `sys.path` itself:

    python3 .agents/skills/qa-run/scripts/qa_tool.py analysis junit report.xml

That line is identical in bash, zsh, PowerShell, and cmd.exe. On Windows, use
`python` if `python3` is not on PATH — that is the only remaining difference.

## Usage

    qa_tool.py analysis    <subcommand> [args]   -> qa_analysis.cli
    qa_tool.py diagnostics <subcommand> [args]   -> qa_diagnostics.cli
    qa_tool.py playwright  <subcommand> [args]   -> playwright_analysis
    qa_tool.py --list                            what this skill bundles

Exit codes pass through from the underlying tool unchanged: 0 success, 1 an
invalid contract, 2 unreadable input or a payload that failed its contract.
"""

import os
import pathlib
import sys

_HERE = pathlib.Path(__file__).resolve().parent
_LIB = _HERE / "lib"

# Tool name -> (module path, callable). Resolved lazily so a skill that bundles
# only part of the toolkit still runs the parts it has.
_TOOLS = {
    "analysis": ("qa_analysis.cli", "main"),
    "diagnostics": ("qa_diagnostics.cli", "main"),
    "playwright": ("playwright_analysis", "main"),
}

_USAGE = """usage: qa_tool.py <tool> <subcommand> [args]

  analysis     parse artifacts, classify errors, validate contracts, diff-guard
  diagnostics  root cause, timeline, priority, repair plans
  playwright   normalize a Playwright report or summarize a trace

  --list       show which tools this skill bundles

examples:
  python3 qa_tool.py analysis junit test-results/results.xml
  python3 qa_tool.py analysis diff-guard change.diff
  python3 qa_tool.py diagnostics report --execution-result qa-artifacts/run.json
  python3 qa_tool.py playwright report test-results/results.json
"""


def _available():
    """Tools whose module is importable from this skill's bundled lib/."""
    found = []
    for name, (module, _) in _TOOLS.items():
        top = module.split(".")[0]
        if (_LIB / top).is_dir() or (_LIB / f"{top}.py").is_file():
            found.append(name)
    return found


def main(argv=None):
    argv = list(sys.argv[1:] if argv is None else argv)

    if not _LIB.is_dir():
        sys.stderr.write(
            f"qa_tool: no bundled library at {_LIB}\n"
            "The skill's deterministic engine is missing. Run: qa repair --project .\n"
        )
        return 2

    # The bundled lib is authoritative: prepend so a same-named package installed
    # elsewhere on the machine cannot shadow the version that shipped with this skill.
    sys.path.insert(0, str(_LIB))

    if not argv or argv[0] in ("-h", "--help", "help"):
        sys.stdout.write(_USAGE)
        return 0

    if argv[0] == "--list":
        available = _available()
        sys.stdout.write(
            "bundled tools: " + (", ".join(available) if available else "none") + "\n"
        )
        return 0 if available else 2

    tool = argv[0]
    if tool not in _TOOLS:
        sys.stderr.write(f"qa_tool: unknown tool {tool!r}\n\n{_USAGE}")
        return 2

    module_name, entry = _TOOLS[tool]
    try:
        module = __import__(module_name, fromlist=[entry])
    except ImportError as exc:
        available = _available()
        sys.stderr.write(
            f"qa_tool: {tool} is not bundled in this skill ({exc}).\n"
            f"This skill bundles: {', '.join(available) if available else 'nothing'}\n"
        )
        return 2

    return getattr(module, entry)(argv[1:])


if __name__ == "__main__":
    # os.environ is untouched: the launcher never mutates the caller's environment.
    del os
    sys.exit(main())
